portalnet/src/socket.rs

```diff
@@ -35,3 +45,85 @@ pub fn stun_for_external(local_socket_addr: &SocketAddr) -> Option<SocketAddr> {
         }
     }
 }
+
+pub fn is_local_addr(external_ip: Option<IpAddr>) -> bool {
+    match external_ip {
+        None => false,
+        Some(external_ip) => {
+            let network_interfaces = local_ip_address::list_afinet_netifas();
+
+            if let Ok(network_interfaces) = network_interfaces {
+                for (_, ip) in network_interfaces.iter() {
+                    if *ip == external_ip {
+                        return true;
+                    }
+                }
+            }
+
+            false
+        }
+    }
+}
+
+pub fn upnp_for_external(listen_addr: SocketAddr) -> Option<SocketAddr> {
+    info!("Connecting to UPnP gateway to map local address to external address");
+    let gateway = match igd_next::search_gateway(Default::default()) {
+        Ok(gateway) => gateway,
+        Err(ref err) => {
+            warn!(error = %err, "Error finding gateway");
+            return None;
+        }
+    };
+
+    let external_ip = match gateway.get_external_ip() {
+        Ok(external_ip) => external_ip,
+        Err(ref err) => {
+            warn!(error = %err, "Error getting external IP");
+            return None;
+        }
+    };
+
+    // Find a local ip to map.
+    // UPnP cannot use an unspecified IP (e.g., 0.0.0.0) to map and thus we need to give a specific IP connected to the gateway.
+    // TODO: Detect the local IP connected to the gateway.
```

Comment:
Yeah, as I recall, this has given us trouble before, that if you randomly select one of the IPs, you'll get the wrong one (for example, if a user is on ethernet or wifi). If you can add this into the current PR, that would be excellent.

Reply:
Thanks for the comment.  I added the feature to find a local IP connected to the gateway and use the IP as the external IP for UPnP NAT.